cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java

```diff
@@ -231,15 +234,31 @@ private Aggregation findAggregation(CubeQuery query) {
 
   private Collection<TimeSeries> convertToQueryResult(CubeQuery query,
                                                       Table<Map<String, String>, Long, Long> aggValuesToTimeValues) {
-
     List<TimeSeries> result = Lists.newArrayList();
+    long count = (query.getEndTs() - query.getStartTs()) / query.getResolution() + 1;
+    if (query.getResolution() == Integer.MAX_VALUE) {
+      count = 1;
+    }
     for (Map.Entry<Map<String, String>, Map<Long, Long>> row : aggValuesToTimeValues.rowMap().entrySet()) {
       List<TimeValue> timeValues = Lists.newArrayList();
       for (Map.Entry<Long, Long> timeValue : row.getValue().entrySet()) {
         timeValues.add(new TimeValue(timeValue.getKey(), timeValue.getValue()));
       }
       Collections.sort(timeValues);
-      result.add(new TimeSeries(query.getMeasureName(), row.getKey(), timeValues));
+      PeekingIterator<TimeValue> timeValueItor = Iterators.peekingIterator(
+        new TimeSeriesInterpolator(timeValues, query.getInterpolator(), query.getResolution()).iterator());
+      long resultTimeStamp = query.getStartTs();
+      List<TimeValue> resultTimeValues = Lists.newArrayList();
+      for (int i = 0; i < count; i++) {
```

Comment:
when the interpolator is null, we might not have values for all the data points and this adds them. as we discussed yesterday, we can have a default interpolator that fills in 0's and in that case we would loop but we the zero's will be supplied by the interpolator.